### PR TITLE
Add tenant access checker

### LINKED ISSUE #13987: Provide a `TenantAccessChecker` API for multi-tenancy
### Description

A `TenantAccessChecker` class should be implemented in the `engine` module. It should be used by `*Processor` classes, or any other classes that process `Command` records created by Client requests. The purpose of the class is to determine if the User making the request has access to the requested Tenant data.

The class should provide the following methods:
* `hasAccess(String tenantId, , List<String> authorizedTenants)` - to determine if a user has access to data from a given tenant. The output can be `Either<Exception, String>` providing the tenantId.
   * If the `tenantId` is listed in the `RecordValueWithTenantPermissions`, the `tenantId` is returned. Otherwise, an error is raised.
* `hasFullAccess(List<String> tenantIds, List<String> authorizedTenants)` - to determine if a user has access to all of the `tenantIds` provided in a list. This will be used later for job polling.

:information_source: We might expand or refactor this class according to different needs as development on the multi-tenancy topic progresses.

### AC
- [x] A `TenantAccessChecker` class
- [x] Test coverage for the `TenantAccessChecker` class

### PR DESCRIPTION
**Pull request by koevskinikola**
## Description

* Add a new Auth module to the Zeebe project that will be responsible
for handling authorization data.
* Add a TenantAccessChecker interface and it's implementation.
The purpose of this class is to provide helper methods for
checking if a provided list of authorized tenants contains
a provided tenant id (or list of ids).
   * This class will be used to validate if a request is authorized
to perform changes on the data of a given tenant.

## Related issues

closes #13987 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: auth/src/main/java/io/camunda/zeebe/auth/impl/TenantAccessCheckerImpl.java
Comment (remcowesterhoud) [inline auth/src/main/java/io/camunda/zeebe/auth/impl/TenantAccessCheckerImpl.java:23]:
```suggestion
    return authorizedTenants.contains(tenantId);
```

🔧

**Review by remcowesterhoud — CHANGES_REQUESTED** [review]
Thanks @koevskinikola 🚀  I have a few suggestions, please have a look 👀

File: auth/src/main/java/io/camunda/zeebe/auth/api/TenantAccessChecker.java
Comment (koevskinikola) [reply auth/src/main/java/io/camunda/zeebe/auth/api/TenantAccessChecker.java:12]: :+1: Makes sense. We have this inconsistency in C7, where the class is called `TenantCommandChecker` but we have `authorizedTenants`. It would be nicer if we align them in C8. I'll apply the changes.

File: auth/pom.xml
Comment (koevskinikola) [reply auth/pom.xml:47]: The plugin is used to generate a `test-jar`. I'm not sure if we would need that for the Auth module. I can see that the plugin is being used for `zeebe-utils`, and it's not very clear to me why.

I'll add it and we can refactor it later in development if needed.

**Comment by koevskinikola** [comment]
@remcowesterhoud I applied your suggestions. Please have another look when you have the time.

**Review by remcowesterhoud — APPROVED** [review]
Great! Thanks for the changes 🚀

**Comment by koevskinikola** [comment]
@korthout let me know if you'd like to review this PR before I merge it.

**Comment by remcowesterhoud** [comment]
> @korthout let me know if you'd like to review this PR before I merge it.

I'd be confident enough to merge this without Nico's approval. There's nothing special in this PR 😄 So if this is blocking other stuff lets just merge it and not wait until Nico has time.

**Comment by koevskinikola** [comment]
bors merge

**Comment by korthout** [comment]
Sorry for the late response, but LGTM too 👍

**Pull request merged**
